pkg/operator/ceph/cluster/watcher.go

```diff
@@ -193,3 +197,83 @@ func (c *clientCluster) getCephCluster() *cephv1.CephCluster {
+	osdNearfullRatio := 85
+
+	var osdsNearfullList []int
+	for _, osdStatus := range osdUsage.OSDNodes {
+		osdId := osdStatus.ID
+		storageUtilization, err := osdStatus.Utilization.Int64()
+		if err == nil {
+			if storageUtilization > int64(osdNearfullRatio) {
+				osdsNearfullList = append(osdsNearfullList, osdId)
+			}
+		}
+
+	}
+	increaseDeviceSetCapacity(cephStatusChecker, osdsNearfullList)
+}
+
+// list the storageClassDeviceSets that are linked to OSDs, and increase there storageSize
+func increaseDeviceSetCapacity(cephStatusChecker *cephStatusChecker, osdsNearfullList []int) {
+	var storageClassDeviceSetsList = util.NewSet()
+	for _, osdID := range osdsNearfullList {
+		storageClassDeviceSetName, err := getDeviceSetName(cephStatusChecker, osdID)
+		if err != nil {
+			logger.Debugf("unable to fetch deviceSet name %+v", err)
+		}
+		storageClassDeviceSetsList.Add(storageClassDeviceSetName)
+	}
+
+	clusterName := cephStatusChecker.clusterInfo.NamespacedName()
+	cephCluster, err := cephStatusChecker.context.RookClientset.CephV1().CephClusters(clusterName.Namespace).Get(context.TODO(), clusterName.Name, metav1.GetOptions{})
+	if err != nil {
+		logger.Errorf("failed to retrieve ceph cluster %q in namespace %q", clusterName.Name, clusterName.Namespace)
+	} else {
+		for storageClassDeviceSet := range storageClassDeviceSetsList.Iter() {
+			for _, storageClassDeviceSetProp := range cephCluster.Spec.Storage.StorageClassDeviceSets {
+				if storageClassDeviceSetProp.Name == storageClassDeviceSet {
+					var defaultDeviceStorage int
+					for _, volumes := range storageClassDeviceSetProp.VolumeClaimTemplates {
+						if volumes.Name == "data" {
+							defaultDeviceStorage = volumes.Spec.Resources.Requests.Storage().Size()
+							break
+						}
+					}
+					storageIncreaseCapacity := defaultDeviceStorage * (1 + (storageClassDeviceSetProp.GrowthPolicy.GrowthRatePercent)/100)
+					storageMaxCapacity := resource.MustParse(storageClassDeviceSetProp.GrowthPolicy.Maxsize)
+					storageMaxCapacitySize := storageMaxCapacity.Size()
+					var increaseStorageValue int
+					if storageMaxCapacitySize > storageIncreaseCapacity {
+						increaseStorageValue = storageIncreaseCapacity
+					} else {
+						increaseStorageValue = storageMaxCapacitySize
+					}
+					logger.Debugf("make StorageClassDeviceSet %q size to %q", storageClassDeviceSet, increaseStorageValue)
+					break
+				}
+			}
+		}
+	}
+}
+
+func getDeviceSetName(cephStatusChecker *cephStatusChecker, osdID int) (string, error) {
+	deploymentName := fmt.Sprintf("rook-ceph-osd-%d", osdID)
+	deployment, err := cephStatusChecker.context.Clientset.AppsV1().Deployments(cephStatusChecker.clusterInfo.Namespace).Get(context.TODO(), deploymentName, metav1.GetOptions{})
+	if err != nil {
+		return "", err
+	} else {
+		if deviceSetName, ok := deployment.GetLabels()[osd.CephDeviceSetLabelKey]; ok {
+			return deviceSetName, nil
+		}
+		return "", fmt.Errorf("unable to get deviceSetName")
```

Comment:
actually we can wrap the error
```suggestion
		return "", errors.Wrap(err, "unable to get deviceSetName")
```